### PR TITLE
Feature/833 Add loki persistence storage to s3 option

### PR DESCRIPTION
**Pull request by trottomv**

### CODE REVIEW

**Review by pauloxnet — CHANGES_REQUESTED** [review]
Separate generic upgrade changes from other specific with the goal of this PR

**Comment by trottomv** [comment]
> Separate generic upgrade changes from other specific with the goal of this PR

We could leave only grafana to the current version which in any case, is updated to a minor version

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf [hunk 1]
```diff
@@ -1,5 +1,16 @@
 locals {
   namespace = kubernetes_namespace_v1.log_storage.metadata[0].name
+
+  s3_storage_enabled = alltrue(
+    [
+      var.s3_access_id != "",
+      local.s3_bucket_name != "",
+      local.s3_host != "",
```
Comment (filippo-20tab) [inline {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf:8]:
```suggestion
      var.s3_bucket_name != "",
      var.s3_host != "",
```

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/variables.tf [hunk 2]
```diff
@@ -13,3 +13,35 @@ variable "grafana_version" {
...
+}
+
+variable "s3_bucket_name" {
+  description = "The S3 bucket name."
+  type        = string
+  default     = ""
+}
+
+variable "s3_host" {
+  description = "The S3 host url."
```
Comment (filippo-20tab) [inline {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/variables.tf:31]:
```suggestion
  description = "The S3 host."
```

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf [hunk 3]
```diff
@@ -37,16 +39,24 @@ resource "helm_release" "loki" {
   namespace  = local.namespace
   repository = "https://grafana.github.io/helm-charts"
   chart      = "loki-stack"
-  version    = "2.6.1"
+  version    = "2.6.4"
+
+  values = local.s3_storage_enabled ? [file("${path.module}/loki/values.yaml")] : []
```
Comment (filippo-20tab) [inline {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf:44]: Let's discuss this. I find this solution is too limiting.

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/routing/main.tf [hunk 4]
```diff
@@ -290,3 +289,87 @@ resource "kubernetes_manifest" "ingressroute_redirect_to_https" {
...
+      namespace = "kube-system"
+    }
+    spec = {
+      basicAuth = {
+        removeHeader = true
+        secret       = kubernetes_secret_v1.metrics_basic_auth[0].metadata[0].name
+      }
+    }
+  }
+}
```
Comment (filippo-20tab) [inline {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/routing/main.tf:328]: These two resources do not have a real reason to exist. Let's change the other two, in order to have them depend on `local.basic_auth_ready`, and mount them according to `local.basic_auth_enabled` too (or not) as needed.

**Review by filippo-20tab — CHANGES_REQUESTED** [review]

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/routing/main.tf (hunk 4 above)
Comment (trottomv) [reply {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/routing/main.tf:328]: The other two "twins" resources are in a different namespace, unfortunately, so it becomes useful to keep them separate and keep this two as is.

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf (hunk 3 above)
Comment (trottomv) [reply {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf:44]: A solution with concatenated values.yaml could be cleaner, I'll try to work on it

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf [hunk 5]
```diff
@@ -37,17 +39,21 @@ resource "helm_release" "loki" {
   namespace  = local.namespace
   repository = "https://grafana.github.io/helm-charts"
   chart      = "loki-stack"
-  version    = "2.6.1"
+  version    = "2.6.4"
+
+  values = concat(
+    [file("${path.module}/loki/values.yaml")], 
+    local.s3_storage_enabled ? [file("${path.module}/loki/s3_storage.yaml")] : [file("${path.module}/loki/pvc_storage.yaml")]
+  )
```
Comment (filippo-20tab) [inline {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf:47]:
```suggestion
  values = [
    file("${path.module}/loki/values.yaml"), 
    file(local.s3_storage_enabled ? "${path.module}/loki/s3_storage.yaml") : "${path.module}/loki/pvc_storage.yaml"))
  ]
```

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/loki/s3_storage.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    storage_config:
+      boltdb_shipper:
+        cache_ttl: 24h # Can be increased for faster performance over longer query periods, uses more disk space
+        shared_store: aws
+      aws:
+        # bucketnames: <YOUR_DO_SPACES_BUCKET_NAME_HERE>
+        # endpoint: <YOUR_DO_SPACES_BUCKET_ENDPOINT_HERE>  # e.g.: nyc1.digitaloceanspaces.com
+        # region: <YOUR_DO_SPACES_BUCKET_REGION_HERE>      # e.g.: nyc1
+        # access_key_id: <YOUR_DO_SPACES_ACCESS_KEY_HERE>
+        # secret_access_key: <YOURDO_SPACES_SECRET_KEY_HERE>
```
Comment (filippo-20tab) [inline {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/loki/s3_storage.yaml:21]: Are these comments really necessary?

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/routing/main.tf [hunk 7]
```diff
@@ -49,7 +48,7 @@ resource "kubernetes_secret_v1" "traefik_basic_auth" {
 }
 
 resource "kubernetes_manifest" "traefik_basic_auth_middleware" {
```
Comment (filippo-20tab) [inline {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/routing/main.tf:50]: I would remove the `-middleware` part from the `metadata.name` attribute of both Middlewares (leaving their Terraform resources name unaltered). There is no need to repeat their type inside their K8s name.

**Review by filippo-20tab — CHANGES_REQUESTED** [review]

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/routing/main.tf (hunk 7 above)
Comment (trottomv) [reply {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/routing/main.tf:50]: ok

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/loki/s3_storage.yaml (hunk 6 above)
Comment (trottomv) [reply {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/loki/s3_storage.yaml]: not really necessary, i can clean this part of yaml

File: {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf (hunk 5 above)
Comment (trottomv) [reply {{cookiecutter.project_dirname}}/terraform/environment/modules/kubernetes/monitoring/main.tf]: ok

**Review by pauloxnet — APPROVED** [review]

**Pull request merged**